### ISSUE TITLE
Feature Request: Eliminate pop-under bugs

### ISSUE THREAD

**Issue by rafpaf** [metabase/metabase#51256]
tl;dr: Pop-unders – where a modal (or some other overlay) unexpectedly shows up underneath another – are a persistent pain point for the embedding team. They're also a source of bugs and tech debt throughout the app.

---

Metabase uses lots of *overlays* - elements that float above the main content of the page. There are quite a few kinds of overlays: modals, popovers, hovercards, toasts, tooltips, and dropdowns. Two overlays can be visible at the same time. This makes "pop-under" bugs possible - where a new overlay opens _underneath_ an existing overlay and so cannot be seen or clicked on.

To prevent pop-under bugs, on master we set z-indexes on various overlays throughout the app. For example:

```ts
// in the LeaveConfirmation modal
<Modal zIndex={5}>

// in the useConfirmation hook
<Modal zIndex={CONFIRM_MODAL_Z_INDEX}>

// in the bulk actions toast styled component
export const BULK_ACTIONS_Z_INDEX = DEFAULT_POPOVER_Z_INDEX + 1;
export const BulkActionsToast = styled.div`
   z-index: ${BULK_ACTIONS_Z_INDEX};
`
```

But giving overlays hard-coded z-indexes makes unexpected interactions possible. For example, if clicking a button in a modal opens a toast, to ensure that the toast is above the modal, we've sometimes assigned the toast component a higher z-index than the modal. But what if, in another part of the app, that toast needs to launch a modal? Then the modal might show up underneath the toast.

We have several bugs like this: #45469, #45468, #49466, #50943, #50971.

Pop-under bugs have been a recurring headache for our embedding team. @oisincoveney told me: "When creating features for the SDK, the timeline is significantly increased when we have to deal with these [pop-under] issues." Since some overlays are incompatible with others, the team had to:
* use a legacy modal for static embedding because the Mantine modal conflicts with legacy popovers,
* delay a key feature in order to convert legacy popovers and dropdowns to the Mantine design system, and
* introduce two distinct portal containers because of the way that popovers and modals interacted.

To solve these issues, they've resorted to ad hoc solutions, like assigning distinct z-indexes.

This approach is not sustainable, for several reasons:
* As mentioned, z-indexes can interact unpredictably.
* These z-indexes are not presently customizable via the SDK. It wouldn't be feasible to let developers customize all of the various z-indexes we use.
* Each bad interaction between overlays must be discovered and solved one by one.
